ksql-api/src/main/java/io/confluent/ksql/api/server/AcksSubscriber.java

```diff
@@ -17,96 +17,107 @@
 
+import io.vertx.core.Context;
 import io.vertx.core.buffer.Buffer;
 import io.vertx.core.http.HttpServerResponse;
 import io.vertx.core.json.JsonObject;
-import java.util.Objects;
-import org.reactivestreams.Subscriber;
 import org.reactivestreams.Subscription;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
- * As the stream of inserts is processed by the back-end, it communicates success (or failure) of
- * each insert by sending a stream of acks back in the other direction. This class is the subscriber
- * which subscribes to that stream of acks and sends them back to the client. It's a reactive
- * streams subscriber so implements back pressure.
+ * A reactive streams subscriber that subscribes to publishers of acks. As it receive acks it writes
+ * them to the HTTP response.
  */
-public class AcksSubscriber implements Subscriber<Void> {
+public class AcksSubscriber extends ReactiveSubscriber<JsonObject> {
 
-  private static final Logger log = LoggerFactory.getLogger(AcksSubscriber.class);
-  private static final int BATCH_SIZE = 4;
   private static final Buffer ACK_RESPONSE_LINE = new JsonObject().put("status", "ok").toBuffer()
       .appendString("\n");
+  private static final Logger log = LoggerFactory.getLogger(AcksSubscriber.class);
+  private static final int REQUEST_BATCH_SIZE = 1000;
 
   private final HttpServerResponse response;
-  private Subscription subscription;
-  private long tokens;
   private Long insertsSent;
   private long acksSent;
+  private boolean drainHandlerSet;
+  private Subscription subscription;
+  private boolean cancelled;
 
-  public AcksSubscriber(final HttpServerResponse response) {
-    this.response = Objects.requireNonNull(response);
+  public AcksSubscriber(final Context context, final HttpServerResponse response) {
+    super(context);
+    this.response = response;
   }
 
   @Override
-  public synchronized void onSubscribe(final Subscription subscription) {
-    Objects.requireNonNull(subscription);
-    if (this.subscription != null) {
-      throw new IllegalStateException("Already subscribed");
-    }
+  protected void afterSubscribe(final Subscription subscription) {
+    makeRequest(REQUEST_BATCH_SIZE);
     this.subscription = subscription;
-    checkRequestTokens();
   }
 
   @Override
-  public synchronized void onNext(final Void vo) {
-    if (tokens == 0) {
-      throw new IllegalStateException("Unsolicited data");
+  public void handleValue(final JsonObject value) {
+    checkContext();
+    if (cancelled) {
+      return;
     }
     response.write(ACK_RESPONSE_LINE);
     acksSent++;
-    tokens--;
     if (insertsSent != null && insertsSent == acksSent) {
       close();
     } else if (response.writeQueueFull()) {
-      response.drainHandler(v -> checkRequestTokens());
+      if (!drainHandlerSet) {
+        response.drainHandler(v -> {
+          drainHandlerSet = false;
+          checkMakeRequest();
+        });
+        drainHandlerSet = true;
```

Comment:
Yep, it will be called on the same thread - this is part of the threading guarantee of Vert.x - handlers set from one context will always be called back on the same context. For standard verticles (like this one), the same context corresponds to the exact same thread (an event loop thread). This means you can code everything in your verticle as single threaded, and not have to worry about synchronized, race conditions, most deadlock etc.